addons/mrp/report/mrp_report_bom_structure.py

```diff
@@ -122,6 +122,41 @@ def _get_report_data(self, bom_id, searchQty=0, searchVariant=False):
             'precision': self.env['decimal.precision'].precision_get('Product Unit of Measure'),
         }
 
+    @api.model
+    def _get_components_closest_forecasted(self, lines, line_quantities, parent_bom, product_info, ignore_stock=False):
+        """
+            Returns a dict mapping a product to the closest date in the forecast report where
+            consummed quantity >= forecasted quantity.
+        """
+        if ignore_stock:
+            return {}
+        closest_forecasted = {}
+        remaining_products = self.env['product.product']
+        product_quantities_info = {}
+        for line in lines:
+            product = line.product_id
+            quantities_info = product_quantities_info.setdefault(product.id, self._get_quantities_info(product, line.product_uom_id, parent_bom, product_info))
+            stock_loc = quantities_info['stock_loc']
+            product_info[product.id]['consumptions'][stock_loc] += line_quantities.get(line.id, 0.0)
+            if product.detailed_type != 'product' or quantities_info['stock_loc'] != 'in_stock':
+                # Use date.min as a sentinel value for _get_stock_availability
+                closest_forecasted[product.id] = date.min
+            elif float_compare(product_info[product.id]['consumptions'][stock_loc], quantities_info['free_qty'], precision_rounding=product.uom_id.rounding) <= 0:
+                closest_forecasted[product.id] = date.min
+            else:
+                remaining_products |= product
+                closest_forecasted[product.id] = date.max
```

Comment:
What if we had the following BoM :
```
- Final - 1
  - Compo A - 3
  - Compo A - 3
```
And we had exactly 3 **Compo A** in stock. Currently, it would display *Available* for the first line (as it's in stock) and the forecast/resupply for the second line.
However, now the first iteration would set `closest_forecasted[product.id]` to `date.min` (aka *Available*), and the second iteration would overwrite it with `date.max` (aka *Unavailable*), and showing for both lines the forecast/resupply that would be fetched there.

Reply:
Fixed it. I added a small unittest to the PR for this specific use case just to be safe.